### ISSUE TITLE
Child Container Not Disposing Services When It Is Disposed

### ISSUE THREAD

**Issue by mickdelaney** [castleproject/Windsor#27]
I'm trying to use a childcontainer to manage a scope for Simple.Web, the expectation is that when I dispose the child container, it should dispose the referenced services. 
But at the moment its not calling dispose on those services. 

**Discussion here:**
https://groups.google.com/forum/?hl=en&fromgroups=#!topic/castle-project-users/ARMLev6jWnQ

**Sample Code here:**
- The DisposesInstances() Test

https://github.com/mickdelaney/Simple.Web/blob/master/src/Simple.Web.Windsor.Tests/HandlerFactoryBuilderTests.cs

**Comment by reijerh** [comment]
It seems the core issue this ticket was about wasn't really fixed, at least the bug is still there in 4.0.

Here's a unit test (adopted from @mickdelaney's fork https://github.com/mickdelaney/Windsor/commit/7e8ab353e701918e6649cfb4bd22e6581ba4755a):

```
[Test]
public void DisposeChildContainerComponentsResolvedAgainstParentContainer()
{
	var parentContainer = new WindsorContainer();

	var childContainer = new WindsorContainer();
	parentContainer.AddChildContainer(childContainer);
	parentContainer.Register(Component.For<DisposableFoo>().LifestyleTransient());

	var first = childContainer.Resolve<DisposableFoo>();
	Assert.IsNotNull(first);
	childContainer.Release(first);
	Assert.AreEqual(1, DisposableFoo.DisposedCount);

	var second = childContainer.Resolve<DisposableFoo>();
	Assert.IsNotNull(second);
	childContainer.Dispose();
	Assert.AreEqual(2, DisposableFoo.DisposedCount);
}
```
Output:
>   Expected: 2
>  But was:  1

(NB: the original unit test registered the component as Singleton, meaning the test would already fail at the first case)

It seems that if we register the `DisposableFoo` at the `childContainer` instead, the object is disposed correctly.

You can see in the debugger that the object is tracked by the `parentContainer` rather than by the `childContainer`.

Kindly requesting moderators to reopen this (don't know who they are currently), @jonorossi ?

**Comment by jonorossi** [comment]
@reijerh this issue is from 2013, please open a new issue and link to this one. I'm not reopening this issue because it would just confuse things with the old history, and especially if we assign a milestone for the fixed version.
